### ISSUE TITLE
Error `protocol error: InternalError: request handler panicked: tree is mutable` when building a crate with a dependency on `bevy 0.14.1` when asking for documentation

### ISSUE THREAD

**Issue by poliorcetics** [rust-lang/rust-analyzer#17914]
**rust-analyzer version**: `07659783fdfd4ec0a0bffa93017e33e31e567e42`

**rustc version**: `rustc 1.80.1 (3f5fd8dd4 2024-08-06)`

**editor or extension**: [`hx`](https://github.com/helix-editor/helix)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`) none of those

**code snippet to reproduce**: 

```bash
cd /tmp
cargo new --bin repro
cd repro
cargo add bevy
hx src/main.rs
```

```rust
fn main() {
    println!("Hello, world!");
}
```

Asking for documentation on `println!` fails with `protocol error: InternalError: request handler panicked: tree is mutable`, which comes from here I think: https://github.com/rust-lang/rust-analyzer/blob/07659783fdfd4ec0a0bffa93017e33e31e567e42/crates/rust-analyzer/src/handlers/dispatch.rs#L268

I haven't found the string `tree is mutable` either in Helix or R-A nor in the vendored dependencies of either

**Comment by ShoyuVanilla** [comment]
I think that it's coming from here

https://github.com/rust-analyzer/rowan/blob/73ed5e7c331c2bedcf9f022e9681fc4784385dda/src/ast.rs#L99

**Comment by lnicola** [comment]
Did we update `rowan`? I ran into that crash in the past when I tried to upgrade, but I didn't realize we're using the new version. It's from https://github.com/rust-analyzer/rowan/pull/151.

**Comment by poliorcetics** [comment]
Ohhh I know what is going on! I installed with `cargo +stable install --git https://github.com/rust-lang/rust-analyzer --branch master --force rust-analyzer` to test and report issues with the latest R-A when I can and the issue is I'm not passing `--locked`. A new version of `rowan` was published a day or so ago and so my last update picked it up.

So the fix is to pass `--locked` when installing in my case, the more general one I don't know. Should I open the issue on `rowan` instead ?

**Comment by lnicola** [comment]
I don't know, it's possible that we're misusing the API and the new check just happens to catch that.

**Comment by ShoyuVanilla** [comment]
> I don't know, it's more likely that we're misusing the API and the new check just happens to catch that.

I agree. Maybe we should check it but I'm working on other issues right now. I'll try looking into this if this is not resolved by the time I can work on this

**Comment by ShoyuVanilla** [comment]
> Ohhh I know what is going on! I installed with `cargo +stable install --git https://github.com/rust-lang/rust-analyzer --branch master --force rust-analyzer` to test and report issues with the latest R-A when I can and the issue is I'm not passing `--locked`. A new version of `rowan` was published a day or so ago and so my last update picked it up.
> 
> So the fix is to pass `--locked` when installing in my case, the more general one I don't know. Should I open the issue on `rowan` instead ?

But is this possible? The r-a is specifying the version as `0.15.15` here,
https://github.com/rust-lang/rust-analyzer/blob/07659783fdfd4ec0a0bffa93017e33e31e567e42/crates/syntax/Cargo.toml#L19

This is quite misterious 🤔

**Comment by lnicola** [comment]
That just means "newer than".

**Comment by ShoyuVanilla** [comment]
Could #15710 be related?

**Comment by lnicola** [comment]
Yeah, that will change how things work, but I don't know when it's going to happen.

**Comment by davidbarsky** [comment]
> Could #15710 be related?

@ShoyuVanilla I've summarized my current understanding of the state/approach in this comment: https://github.com/rust-lang/rust-analyzer/issues/15710#issuecomment-2294892785

**Comment by ShoyuVanilla** [comment]
> > Could #15710 be related?
> 
> @ShoyuVanilla I've summarized my current understanding of the state/approach in this comment: [#15710 (comment)](https://github.com/rust-lang/rust-analyzer/issues/15710#issuecomment-2294892785)

Cool! That makes clear the current status and what should be done.
I'll check if there is anything I can help it when I finish #17814

**Comment by Veykril** [comment]
Worst case we can pin the current version we have in the toml file for now. You'll only run into this if the lock file gets bumped (or if you install it like OP did)
